lib/Dialect/FIRRTL/Transforms/GrandCentral.cpp

```diff
@@ -0,0 +1,499 @@
+
+using namespace circt;
+using namespace firrtl;
+
+//===----------------------------------------------------------------------===//
+// Pass Implementation
+//===----------------------------------------------------------------------===//
+
+namespace {
+/// Mutable store of information about an Element in an interface.  This is
+/// derived from information stored in the "elements" field of an
+/// "AugmentedBundleType".  This is updated as more information is known about
+/// an Element.
+struct ElementInfo {
+  /// Encodes the "tpe" of an element.  This is called "Kind" to avoid
+  /// overloading the meeaning of "Type" (which also conflicts with mlir::Type).
+  enum Kind {
+    Error = -1,
+    Ground,
+    Vector,
+    Bundle,
+    String,
+    Boolean,
+    Integer,
+    Double
+  };
+  /// The "tpe" field indicating if this element of the interface is a ground
+  /// type, a vector type, or a bundle type.  Bundle types are nested
+  /// interfaces.
+  Kind tpe;
+  /// A string description that will show up as a comment in the output Verilog.
+  StringRef description;
+  /// The width of this interface.  (This is only non-negative for ground or
+  /// vector types.)
+  int32_t width = -1;
+  /// The depth of the interface.  (This is one for ground types and greater
+  /// than one for vector types.)
+  uint32_t depth = 0;
+  /// Indicate if this element was found in the circuit.
+  bool found = false;
+  /// Trakcs location information about what was used to build this element.
+  SmallVector<Location> locations = SmallVector<Location>();
+  /// True if this is a ground or vector type and it was not (statefully) found.
+  /// This indicates that an interface element, which is composed of ground and
+  /// vector types, found no matching, annotated components in the circuit.
+  bool isMissing() { return !found && (tpe == Ground || tpe == Vector); }
+};
+
+/// Stores a decoded Grand Central AugmentedField
+struct AugmentedField {
+  /// The name of the field.
+  StringRef name;
+  /// An optional descripton that the user provided for the field.  This should
+  /// become a comment in the Verilog.
+  StringRef description;
+  /// The "type" of the field.
+  ElementInfo::Kind tpe;
+};
+
+/// Stores a decoded Grand Central AugmentedBundleType.
+struct AugmentedBundleType {
+  /// The name of the interface.
+  StringRef defName;
+  /// The elements that make up the body of the interface.
+  SmallVector<AugmentedField> elements;
+};
+
+/// Convert an arbitrary attributes into an optional AugmentedField.  Returns
+/// None if the attribute is an invalid AugmentedField.
+static Optional<AugmentedField> decodeField(Attribute maybeField) {
+  auto field = maybeField.dyn_cast_or_null<DictionaryAttr>();
+  if (!field)
+    return {};
+  auto tpeString = field.getAs<StringAttr>("tpe");
+  auto name = field.getAs<StringAttr>("name");
+  if (!name || !tpeString)
+    return {};
+  auto tpe = llvm::StringSwitch<ElementInfo::Kind>(tpeString.getValue())
```

Comment:
Nice! Wasn't aware of `StringSwitch` :sunglasses:

Reply:
Yeah, it's come in handy before. 😄